resources/fetchersManager/data.go

```diff
@@ -23,75 +23,62 @@ import (
 	"sync"
 	"time"
 
-	"github.com/elastic/beats/v7/libbeat/common/atomic"
 	"github.com/elastic/cloudbeat/resources/fetching"
 	"github.com/elastic/elastic-agent-libs/logp"
 )
 
 // Data maintains a cache that is updated by Fetcher implementations registered
 // against it. It sends the cache to an output channel at the defined interval.
 type Data struct {
-	log        *logp.Logger
-	timeout    time.Duration
-	interval   time.Duration
-	fetchers   FetchersRegistry
-	stop       chan struct{}
-	stopNotice chan time.Duration
+	log *logp.Logger
+
+	// Duration of a single fetcher
+	timeout time.Duration
+
+	// Duration between two consecutive cycles
+	interval time.Duration
+
+	fetchers FetchersRegistry
+
+	ctx    context.Context
+	cancel context.CancelFunc
 }
 
-type Stop func(context.Context, time.Duration)
+func NewData(ctx context.Context, log *logp.Logger, interval time.Duration, timeout time.Duration, fetchers FetchersRegistry) (*Data, error) {
+	ctx, cancel := context.WithCancel(ctx)
 
-// NewData returns a new Data instance.
-// interval is the duration the manager wait between two consecutive cycles.
-// timeout is the maximum duration the manager wait for a single fetcher to return results.
-func NewData(log *logp.Logger, interval time.Duration, timeout time.Duration, fetchers FetchersRegistry) (*Data, error) {
 	return &Data{
-		log:        log,
-		timeout:    timeout,
-		interval:   interval,
-		fetchers:   fetchers,
-		stop:       make(chan struct{}),
-		stopNotice: make(chan time.Duration),
+		log:      log,
+		timeout:  timeout,
+		interval: interval,
+		fetchers: fetchers,
+		ctx:      ctx,
+		cancel:   cancel,
 	}, nil
 }
 
 // Run starts all configured fetchers to collect resources.
-func (d *Data) Run(ctx context.Context) Stop {
-	go d.fetchAndSleep(ctx)
-	called := atomic.NewBool(false)
-	return func(ctx context.Context, grace time.Duration) {
-		defer called.Store(true)
-		if called.Load() {
-			return
-		}
-		d.stopData(ctx, grace)
-	}
+func (d *Data) Run() {
+	go d.fetchAndSleep(d.ctx)
+}
+
+func (d *Data) Stop() {
```

Comment:
What happen if you call this function twice?

Reply:
as discussed, this seems to work fine when called twice (no panics), but as per the beats docs, it will only be called [once](https://github.com/elastic/beats/blob/c4398dcf5f62ca4606eaf7d1dbf76635ecf76146/libbeat/beat/beat.go#L49-L510)